server/src/proxy/grpc/write.rs

```diff
@@ -152,18 +195,128 @@ impl<Q: QueryExecutor + 'static> Proxy<Q> {
             .await?;
         }
 
-        let resp = WriteResponse {
+        Ok(WriteResponse {
             success: success as u32,
             header: Some(build_ok_header()),
             ..Default::default()
+        })
+    }
+
+    async fn maybe_forward_write(
+        &self,
+        req: WriteRequest,
+    ) -> Result<(WriteRequest, Vec<BoxFuture<Result<WriteResponse>>>)> {
+        // Split write request into multiple requests, each request contains table
+        // belong to one remote engine.
+        let tables = req
+            .table_requests
+            .iter()
+            .map(|table_request| table_request.table.clone())
+            .collect();
+
+        let route_data = self
+            .router
+            .route(RouteRequest {
+                context: req.context.clone(),
+                tables,
+            })
+            .await?;
+        let forwarded_table_route = route_data
+            .into_iter()
+            .filter_map(|router| {
+                router
+                    .endpoint
+                    .map(|endpoint| (router.table, endpoint.into()))
+            })
+            .filter(|router| !self.forwarder.is_local_endpoint(&router.1))
+            .collect::<HashMap<_, _>>();
+
+        // No table need to be forwarded.
+        if forwarded_table_route.is_empty() {
+            return Ok((req, vec![]));
+        }
+
+        let mut table_requests_to_local = WriteRequest {
+            table_requests: Vec::with_capacity(req.table_requests.len()),
```

Comment:
The capacity may be able to be figured out here: `req.table_requests.len() - forward_table_route.len()`. And note for overflow after that.